protocol/validation/validation.go

```diff
@@ -145,133 +147,154 @@ func checkValid(vs *validationState, e bc.Entry) error {
+		if *e.Witness.AssetDefinition.InitialBlockId != vs.blockchainID {
+			return errors.WithDetailf(errWrongBlockchain, "current blockchain %x, asset defined on blockchain %x", vs.blockchainID.Bytes(), e.Witness.AssetDefinition.InitialBlockId.Bytes())
 		}
 
 		computedAssetID := e.Witness.AssetDefinition.ComputeAssetID()
-		if computedAssetID != e.Body.Value.AssetID {
-			return errors.WithDetailf(errMismatchedAssetID, "asset ID is %x, issuance wants %x", computedAssetID.Bytes(), e.Body.Value.AssetID.Bytes())
+		if computedAssetID != *e.Body.Value.AssetId {
+			return errors.WithDetailf(errMismatchedAssetID, "asset ID is %x, issuance wants %x", computedAssetID.Bytes(), e.Body.Value.AssetId.Bytes())
 		}
 
 		err := vm.Verify(NewTxVMContext(vs.tx, e, e.Witness.AssetDefinition.IssuanceProgram, e.Witness.Arguments))
 		if err != nil {
 			return errors.Wrap(err, "checking issuance program")
 		}
 
-		var anchored bc.Hash
-		switch a := e.Anchor.(type) {
+		anchor, ok := vs.tx.Entries[*e.Body.AnchorId]
+		if !ok {
+			return errors.Wrapf(errMissingEntry, "entry for issuance anchor %x not found", e.Body.AnchorId.Bytes())
+		}
+
+		var anchored *bc.Hash
+		switch a := anchor.(type) {
 		case *bc.Nonce:
-			anchored = a.Witness.AnchoredID
+			anchored = a.Witness.AnchoredId
 
 		case *bc.Spend:
-			anchored = a.Witness.AnchoredID
+			anchored = a.Witness.AnchoredId
 
 		case *bc.Issuance:
-			anchored = a.Witness.AnchoredID
+			anchored = a.Witness.AnchoredId
 
 		default:
-			return errors.WithDetailf(errEntryType, "issuance anchor has type %T, should be nonce, spend, or issuance", e.Anchor)
+			return errors.WithDetailf(errEntryType, "issuance anchor has type %T, should be nonce, spend, or issuance", anchor)
 		}
 
-		if anchored != vs.entryID {
+		if *anchored != vs.entryID {
 			return errors.WithDetailf(errMismatchedReference, "issuance %x anchor is for %x", vs.entryID.Bytes(), anchored.Bytes())
 		}
 
 		anchorVS := *vs
-		anchorVS.entryID = e.Body.AnchorID
-		err = checkValid(&anchorVS, e.Anchor)
+		anchorVS.entryID = *e.Body.AnchorId
+		err = checkValid(&anchorVS, anchor)
 		if err != nil {
 			return errors.Wrap(err, "checking issuance anchor")
 		}
 
 		destVS := *vs
 		destVS.destPos = 0
-		err = checkValidDest(&destVS, &e.Witness.Destination)
+		err = checkValidDest(&destVS, e.Witness.Destination)
 		if err != nil {
 			return errors.Wrap(err, "checking issuance destination")
 		}
 
-		if vs.tx.Body.Version == 1 && e.Body.ExtHash != (bc.Hash{}) {
+		if vs.tx.Body.Version == 1 && !e.Body.ExtHash.IsZero() {
 			return errNonemptyExtHash
 		}
 
 	case *bc.Spend:
-		err := vm.Verify(NewTxVMContext(vs.tx, e, e.SpentOutput.Body.ControlProgram, e.Witness.Arguments))
+		spentOutputEntry, ok := vs.tx.Entries[*e.Body.SpentOutputId]
+		if !ok {
+			return errors.Wrapf(errMissingEntry, "entry for prevout %x not found", e.Body.SpentOutputId.Bytes())
+		}
+		spentOutput, ok := spentOutputEntry.(*bc.Output)
```

Comment:
What do you think of https://github.com/chain/chain/pull/887/commits/adc6dd38801fe507a1daccb19bcecf9c4da100ca ?